app/services/order_purchaser.rb

```diff
@@ -0,0 +1,81 @@
+class OrderPurchaser
+  attr_reader :order, :params, :user
+
+  delegate :order_notification_recipient, to: :order
+  delegate :quantities_changed?, to: :order_detail_updater
+
+  def initialize(acting_as:, order:, order_in_past:, params:, user:)
+    @acting_as = acting_as
+    @order = order
+    @order_in_past = order_in_past
+    @params = params
+    @user = user
+  end
+
+  def purchase!
+    order.transaction do
+      return if order_detail_updater.update! && quantities_changed?
```

Comment:
The name `QuantityUpdateChangeException` was misleading because it was wrong. A quantity update should update `order_detail` but not move the `order` to `purchased`, but not raise an exception. I renamed the exception (which I'd still like to phase out).